logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/composite/SendButtonMessageUseCase.kt

```diff
@@ -0,0 +1,122 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.feature.message.composite
+
+import com.benasher44.uuid.uuid4
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.id.CurrentClientIdProvider
+import com.wire.kalium.logic.data.id.QualifiedID
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageContent
+import com.wire.kalium.logic.data.message.PersistMessageUseCase
+import com.wire.kalium.logic.data.message.mention.MessageMention
+import com.wire.kalium.logic.data.properties.UserPropertyRepository
+import com.wire.kalium.logic.data.sync.SlowSyncRepository
+import com.wire.kalium.logic.data.sync.SlowSyncStatus
+import com.wire.kalium.logic.feature.message.MessageSendFailureHandler
+import com.wire.kalium.logic.feature.message.MessageSender
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.onFailure
+import com.wire.kalium.util.DateTimeUtil
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.first
+
+@Suppress("LongParameterList")
+/**
+ * @sample samples.logic.MessageUseCases.sendingBasicTextMessage
+ * @sample samples.logic.MessageUseCases.sendingTextMessageWithMentions
+ */
+class SendButtonMessageUseCase internal constructor(
+    private val persistMessage: PersistMessageUseCase,
+    private val selfUserId: QualifiedID,
+    private val provideClientId: CurrentClientIdProvider,
+    private val slowSyncRepository: SlowSyncRepository,
+    private val messageSender: MessageSender,
+    private val messageSendFailureHandler: MessageSendFailureHandler,
+    private val userPropertyRepository: UserPropertyRepository,
+    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl,
+    private val scope: CoroutineScope
+) {
+
+    suspend operator fun invoke(
+        conversationId: ConversationId,
+        text: String,
+        mentions: List<MessageMention> = emptyList(),
+        quotedMessageId: String? = null,
+        buttons: List<String> = listOf()
+    ): Either<CoreFailure, Unit> = scope.async(dispatchers.io) {
```

Comment:
```suggestion
    ): Either<CoreFailure, Unit> = withContext(dispatchers.io) {
```

Question: is there any special need, why we need to pass the `coroutineScope` to the UseCase ? Usually we are just doing as suggested ☝️  , just because I guess Use cases are exposed to clients, and I don't know how that will look in other platforms (food for thought) not blocker, just about uniformity

Reply:
this mean the message will be sent async like we do in text message 
1. send a message 
2. close navigate back before it is done sending 
3. message will be sent anyway